lib/vtls/sectransp.c

```diff
@@ -1386,6 +1479,228 @@ set_ssl_version_min_max(struct Curl_easy *data, struct connectdata *conn,
+  }
+  for(i = 0UL ; i < all_ciphers_count ; i++) {
+#if CURL_BUILD_MAC
+   /* There's a known bug in early versions of Mountain Lion where ST's ECC
+      ciphers (cipher suite 0xC001 through 0xC032) simply do not work.
+      Work around the problem here by disabling those ciphers if we are
+      running in an affected version of OS X. */
+    if(darwinver_maj == 12 && darwinver_min <= 3 &&
+       all_ciphers[i] >= 0xC001 && all_ciphers[i] <= 0xC032) {
+      continue;
+    }
+#endif /* CURL_BUILD_MAC */
+    if(is_cipher_suite_strong(all_ciphers[i])) {
+      allowed_ciphers[allowed_ciphers_count++] = all_ciphers[i];
+    }
+  }
+  err = SSLSetEnabledCiphers(backend->ssl_ctx, allowed_ciphers,
+                             allowed_ciphers_count);
+  Curl_safefree(all_ciphers);
+  Curl_safefree(allowed_ciphers);
+  if(err != noErr) {
+    failf(data, "SSL: SSLSetEnabledCiphers() failed: OSStatus %d", err);
+    return CURLE_SSL_CIPHER;
+  }
+  return CURLE_OK;
+}
+
+static CURLcode sectransp_set_selected_ciphers(struct connectdata *conn,
+                                               int sockindex,
+                                               const char *ciphers)
+{
+  struct Curl_easy *data = conn->data;
+  struct ssl_connect_data *connssl = &conn->ssl[sockindex];
+  struct ssl_backend_data *backend = connssl->backend;
+  size_t ciphers_count = 0;
+  const char *cipher_start = ciphers;
+  OSStatus err = noErr;
+  SSLCipherSuite selected_ciphers[NUM_OF_CIPHERS];
+
+  if(!ciphers)
+    return CURLE_OK;
+
+  while(is_separator(*ciphers))     /* Skip initial separators. */
+    ciphers++;
+  if(!*ciphers)
+    return CURLE_OK;
+
+  cipher_start = ciphers;
+  while(*cipher_start && ciphers_count < NUM_OF_CIPHERS) {
+    bool cipher_found = FALSE;
+    size_t cipher_len = 0;
+    const char *cipher_end = NULL;
+    bool tls_name = FALSE;
+    bool ssl_name = FALSE;
+
+    /* Skip separators */
+    while(is_separator(*cipher_start))
+       cipher_start++;
+    if(*cipher_start == '\0') {
+      break;
+    }
+    /* Find last position of a cipher in the ciphers string */
+    cipher_end = cipher_start;
+    while (*cipher_end != '\0' && !is_separator(*cipher_end)) {
+      ++cipher_end;
+    }
+
+    /* In the future we can add support of cipher groups like cipher suites
+       supported by TLS1.0 or TLS1.3 etc. It would be subset of similar lists
+       or keywords supprted by OpenSSL. We do not need to implement complicated
+       predicates like +-! signs etc. We can support just certain group names
+       and marker character for a group name would be 4th symbol in a name:
+       all IANA cipher names contain underscore '_' symbol there after a
+       protocol name (SSL_ or TLS_), if the 4th symbol is different then we
+       need to treat it as an alias name or a group name.
+       If it matches to a group name then we fill our array with all ciphers
+       in the group. This is a good place we can branch those two cases.
```

Comment:
That should be "a good place where we can branch those two cases".

Reply:
After some time and thought iterations, I reconsidered the point about cipher groups.
It requires more maintenance in the future of the source code and documentation, and doesn't give much advantage, from my point of view. I think I better remove that comment. If in the future we'll need to add that feature, we'll find place to add the new code.
The main purpose of that comment on my side was to discuss that feature or think more about it. I did the latter and made some decision.

Another point for me to think a bit more was to keep or remove SSL names from the code.
Existing code maps cipher number to the name only for logging. Also, first ciphers are mapped to 2 names: TLS_ name and SSL_ name. That is only for logging, and I looked at the OpenSSL code. They use only TLS names for logging (openssl-1.1.1h/ssl/t1_trce.c) and probably for some purposes, I did not find SSL_ names, only TLS_.
I think we also can drop SSL_ names. On the other hand, if we drop them, we break compatibility with existing version. I doubt someone relies on cipher names in logs for some automation or text matching, but I'd like to know your opinion.
I'll update PR with 2 versions, one addressing your comments, and another will remove SSL names on top of it.
Please take a look, when they'll be ready and let me know if you are ok to remove SSL names.

Thank you.